sklearn/model_selection/_split.py

```diff
@@ -2787,6 +2789,14 @@ def train_test_split(
 
         train, test = next(cv.split(X=arrays[0], y=stratify))
 
+    xp, is_array_api_compliant = get_namespace(arrays[0])
+    if is_array_api_compliant:
+        # For indexing, the data and index arrays need to be of the same type
+        # and on the same device.
+        _device = array_api_device(arrays[0])
+        train = xp.asarray(train, device=_device)
+        test = xp.asarray(test, device=_device)
```

Comment:
For the record, this change is very similar the one from concurrent PR: #28407. The difference being the extra `not _is_numpy_namespace(xp)` in the condition.

Not sure what we really want but we can merge this first and refine with extra tests with an updated version of #28407.

Reply:
Happy with whatever you and @betatim prefer here. This is needed to support testing array API support against numpy main/2.0.0rc1 soon, but it doesn't look like anything else will go wrong if the numpy changes get merged before this PR.

Reply:
I'd vote to merge this PR with the `if` as it is. I think the reason I added the `and not _is_numpy_namespace(xp)` was to avoid going down this branch if the arrays are Numpy arrays. But I think that it is harmless to take this branch also for Numpy arrays? So if less tests fail with this than right now, lets go with it. We can always course correct in #28407, which I think will only be merged after the `r2_score` PR, which I guess is going to be merged after his PR?

(and it kinda annoys me to see lots of `if array_api and not is_numpy` in the code.)